presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,231 @@
+            throw new PrestoException(NATIVEPLANCHECKER_CONNECTION_ERROR, "I/O error getting native plan checker response", e);
+        }
+        finally {
+            LOG.debug("Native plan validation complete [fragment: %s, root: %s]", planFragment.getId(), planFragment.getRoot().getId());
+        }
+    }
+
+    private Request buildRequest(String requestBodyJson)
+    {
+        // Use native sidecar plan conversion endpoint to validate
+        String planConversionUrl = nodeManager.getSidecarNode().getHttpUri().toString() + PLAN_CONVERSION_ENDPOINT;
+
+        Request.Builder builder = new Request.Builder()
+                .url(planConversionUrl)
+                .addHeader("CONTENT_TYPE", "APPLICATION_JSON")
+                .post(RequestBody.create(JSON_CONTENT_TYPE, requestBodyJson));
+
+        return builder.build();
+    }
+
+    private PlanConversionFailure processResponseFailure(Response response) throws IOException
+    {
+        if (response.body() == null) {
+            throw new PrestoException(NATIVEPLANCHECKER_UNKNOWN_CONVERSION_FAILURE, "Error response without failure from native plan checker with code: " + response.code());
+        }
+
+        PlanConversionResponse planConversionResponse = PLAN_CONVERSION_RESPONSE_JSON_CODEC.fromJson(response.body().bytes());
+        if (planConversionResponse.getFailures().isEmpty()) {
+            throw new PrestoException(NATIVEPLANCHECKER_UNKNOWN_CONVERSION_FAILURE, "Error response without failure from native plan checker with code: " + response.code());
+        }
+
+        return planConversionResponse.getFailures().get(0);
+    }
+
+    private static class CheckInternalVisitor
+            extends PlanVisitor<Boolean, Void>
+    {
+        @Override
+        public Boolean visitTableScan(TableScanNode tableScan, Void context)
+        {
+            TableHandle handle = tableScan.getTable();
+            return ConnectorId.isInternalSystemConnector(handle.getConnectorId());
+        }
+
+        @Override
+        public Boolean visitPlan(PlanNode node, Void context)
+        {
+            for (PlanNode child : node.getSources()) {
+                if (child.accept(this, context)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    public static class PlanConversionResponse
+    {
+        private final List<PlanConversionFailure> failures;
+
+        @JsonCreator
+        public PlanConversionResponse(
+                @JsonProperty("failures") List<PlanConversionFailure> failures)
+        {
+            this.failures = ImmutableList.copyOf(requireNonNull(failures, "failures is null"));
+        }
+
+        @JsonProperty
+        public List<PlanConversionFailure> getFailures()
+        {
+            return failures;
+        }
+    }
+
+    public static class PlanConversionFailure
+            implements ErrorCodeSupplier
+    {
+        private final String type;
```

Comment:
What is this "type", and how is it different from Errorcode.type?

Reply:
This `PlanConversionFailure` was a simple copy of `ExecutionFailureInfo` and I'll be removing it and explain below. 

But "type" is really the classname of the exception